modules/core/src/matrix.cpp

```diff
@@ -4257,6 +4257,64 @@ Mat Mat::reshape(int _cn, int _newndims, const int* _newsz) const
             return reshape(_cn, _newsz[0]);
     }
 
+    if (_newsz && _newndims > 0 && isContinuous())
+    {
+        CV_Assert(_cn >= 0);
+
+        size_t total_elem1_ref = this->total() * this->channels();
+        size_t total_elem1 = (_cn > 0) ? _cn : 1;
```

Comment:
Use same parameter behaviour:
`1` --> `this->channels()`
https://github.com/Itseez/opencv/blob/47e7a7a7ce11884ddaf46b0f7aa7774ee7708915/modules/core/src/matrix.cpp#L970-L971

+fix/add test for this case